ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafAggregator.java

```diff
@@ -50,16 +51,35 @@ public GenericRow apply(final Struct k, final GenericRow rowValue, final Generic
     // compute the aggregation and write it into the output row. Its assumed that
     // the columns written by this statement do not overlap with those written by
     // the above statement.
-    aggValToAggFunctionMap.forEach((key, value) ->
+    aggValToAggFunctionMap.forEach((functionIndex, ksqlAggregateFunction) ->
         aggRowValue.getColumns().set(
-            key,
-            value.aggregate(
-                rowValue.getColumns().get(value.getArgIndexInValue()),
-                aggRowValue.getColumns().get(key))));
+            functionIndex,
+            ksqlAggregateFunction.aggregate(
+                rowValue.getColumns().get(ksqlAggregateFunction.getArgIndexInValue()),
+                aggRowValue.getColumns().get(functionIndex))));
 
     return aggRowValue;
   }
 
+  @SuppressWarnings("unchecked")
+  public ValueMapper<GenericRow, GenericRow> getMapper() {
+
+    return (aggRow) -> {
+      final List<Object> columns = Stream.generate(String::new).limit(aggRow.getColumns().size())
+          .collect(Collectors.toList());
+
+      for (int idx = 0; idx < nonFuncColumnCount; idx++) {
+        columns.set(idx, aggRow.getColumns().get(idx));
+      }
+
+      aggValToAggFunctionMap.forEach((functionIndex, ksqlAggregateFunction) ->
```

Comment:
my preference is to avoid `.forEach` when you can easily do a for each loop. I think it has unnecessary allocations, doesn't make the code easier to read and makes debugging annoying (stepping through the loop is much harder). Andy disagrees with me, but I hope you can come around on this one! Feel free to drop if you disagree